# Guide to cleaning and preparing data in python

Real world data is dirty. In fact, around 80% of a data scientist's
time is spent collecting, cleaning and Preparing data.these tedious
but necessary steps make the data suitable for any model we want to
and ensure the high quality of data.

This might be tricky sometimes,so in this github, i would like to 
make these processes easier by showing some methods and functions 
used to clean and prepare the data.

Here we are using a netflix dataset available on kaggle

### 1. Quick Dataset Overview

In [6]:
import pandas as pd
file = 'D:/TUTORIALS/Datas/netflix_titles.csv'
df_netflix_2019 = pd.read_csv(file)
df_netflix_2019.dtypes

show_id         object
type            object
title           object
director        object
cast            object
country         object
date_added      object
release_year     int64
rating          object
duration        object
listed_in       object
description     object
dtype: object

In [5]:
df_netflix_2019.shape 

(8807, 12)

This dataset contains 8807 rows and 12 columns


### 2. Identify missing data 

Missing data sometimes occurs when data collection was done improperly, mistakes were made in data entry, or data values were not stored. This happens often, and we should know how to identify it.

Create a percentage list with .isnull()

In [9]:
df_netflix_2019.isnull().sum().sort_values(ascending=False)

director        2634
country          831
cast             825
date_added        10
rating             4
duration           3
show_id            0
type               0
title              0
release_year       0
listed_in          0
description        0
dtype: int64

That being said, I usually represent the missing values in percentages, so I have a clearer picture of the missing data. The following code shows the above output in %

In [10]:
# % of rows missing in each column

for column in df_netflix_2019.columns:
    percentage = df_netflix_2019[column].isnull().mean()
    print(f'{column}: {round(percentage*100, 2)}%')

show_id: 0.0%
type: 0.0%
title: 0.0%
director: 29.91%
cast: 9.37%
country: 9.44%
date_added: 0.11%
release_year: 0.0%
rating: 0.05%
duration: 0.03%
listed_in: 0.0%
description: 0.0%


### 3. Dealing with missing data

There are different ways of dealing with missing data.the correct approach to handling missing data will be highly influenced by the data and goals your project has
The following cover 3 simple ways of dealing with missing data

#### Remove a column or row with .drop, .dropna, or .isnull

If you consider it’s necessary to remove a column because it has
too many empty rows, you can use .drop() and add axis=1 as a
parameter to indicate that what you want to drop is a column.



In [20]:
#Drop column
df_netflix_2019.drop('director',axis=1)

show_id     type                  title  \
0         s1    Movie   Dick Johnson Is Dead   
1         s2  TV Show          Blood & Water   
2         s3  TV Show              Ganglands   
3         s4  TV Show  Jailbirds New Orleans   
4         s5  TV Show           Kota Factory   
...      ...      ...                    ...   
8802   s8803    Movie                 Zodiac   
8803   s8804  TV Show            Zombie Dumb   
8804   s8805    Movie             Zombieland   
8805   s8806    Movie                   Zoom   
8806   s8807    Movie                 Zubaan   

                                                   cast        country  \
0                                                   NaN  United States   
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2     Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                   NaN            NaN   
4     Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   
...                                                 ...            ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...  United States   
8803                                                NaN            NaN   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...  United States   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   

              date_added  release_year rating   duration  \
0     September 25, 2021          2020  PG-13     90 min   
1     September 24, 2021          2021  TV-MA  2 Seasons   
2     September 24, 2021          2021  TV-MA   1 Season   
3     September 24, 2021          2021  TV-MA   1 Season   
4     September 24, 2021          2021  TV-MA  2 Seasons   
...                  ...           ...    ...        ...   
8802   November 20, 2019          2007      R    158 min   
8803        July 1, 2019          2018  TV-Y7  2 Seasons   
8804    November 1, 2019          2009      R     88 min   
8805    January 11, 2020          2006     PG     88 min   
8806       March 2, 2019          2015  TV-14    111 min   

                                              listed_in  \
0                                         Documentaries   
1       International TV Shows, TV Dramas, TV Mysteries   
2     Crime TV Shows, International TV Shows, TV Act...   
3                                Docuseries, Reality TV   
4     International TV Shows, Romantic TV Shows, TV ...   
...                                                 ...   
8802                     Cult Movies, Dramas, Thrillers   
8803             Kids' TV, Korean TV Shows, TV Comedies   
8804                            Comedies, Horror Movies   
8805                 Children & Family Movies, Comedies   
8806     Dramas, International Movies, Music & Musicals   

                                            description  
0     As her father nears the end of his life, filmm...  
1     After crossing paths at a party, a Cape Town t...  
2     To protect his family from a powerful drug lor...  
3     Feuds, flirtations and toilet talk go down amo...  
4     In a city of coaching centers known to train I...  
...                                                 ...  
8802  A political cartoonist, a crime reporter and a...  
8803  While living alone in a spooky town, a young g...  
8804  Looking to survive in a world taken over by zo...  
8805  Dragged from civilian life, a former superhero...  
8806  A scrappy but poor boy worms his way into a ty...  

[8807 rows x 11 columns]

However,most of the time is just enough to remove the rows containing those empty values.there are different ways to do so.

In [13]:
#Drop row
no_director = df_netflix_2019[df_netflix_2019['director'].isnull()].index
df_netflix_2019.drop(no_director, axis=0)

#~ + .isnull()
df_netflix_2019[~df_netflix_2019['director'].isnull()]

#dropna()
df_netflix_2019.dropna(subset=['director'])


show_id     type                             title  \
0         s1    Movie              Dick Johnson Is Dead   
2         s3  TV Show                         Ganglands   
5         s6  TV Show                     Midnight Mass   
6         s7    Movie  My Little Pony: A New Generation   
7         s8    Movie                           Sankofa   
...      ...      ...                               ...   
8801   s8802    Movie                           Zinzana   
8802   s8803    Movie                            Zodiac   
8804   s8805    Movie                        Zombieland   
8805   s8806    Movie                              Zoom   
8806   s8807    Movie                            Zubaan   

                           director  \
0                   Kirsten Johnson   
2                   Julien Leclercq   
5                     Mike Flanagan   
6     Robert Cullen, José Luis Ucha   
7                      Haile Gerima   
...                             ...   
8801                Majid Al Ansari   
8802                  David Fincher   
8804                Ruben Fleischer   
8805                   Peter Hewitt   
8806                    Mozez Singh   

                                                   cast  \
0                                                   NaN   
2     Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...   
5     Kate Siegel, Zach Gilford, Hamish Linklater, H...   
6     Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
7     Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
...                                                 ...   
8801  Ali Suliman, Saleh Bakri, Yasa, Ali Al-Jabri, ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...   

                                                country          date_added  \
0                                         United States  September 25, 2021   
2                                                   NaN  September 24, 2021   
5                                                   NaN  September 24, 2021   
6                                                   NaN  September 24, 2021   
7     United States, Ghana, Burkina Faso, United Kin...  September 24, 2021   
...                                                 ...                 ...   
8801                       United Arab Emirates, Jordan       March 9, 2016   
8802                                      United States   November 20, 2019   
8804                                      United States    November 1, 2019   
8805                                      United States    January 11, 2020   
8806                                              India       March 2, 2019   

      release_year rating  duration  \
0             2020  PG-13    90 min   
2             2021  TV-MA  1 Season   
5             2021  TV-MA  1 Season   
6             2021     PG    91 min   
7             1993  TV-MA   125 min   
...            ...    ...       ...   
8801          2015  TV-MA    96 min   
8802          2007      R   158 min   
8804          2009      R    88 min   
8805          2006     PG    88 min   
8806          2015  TV-14   111 min   

                                              listed_in  \
0                                         Documentaries   
2     Crime TV Shows, International TV Shows, TV Act...   
5                    TV Dramas, TV Horror, TV Mysteries   
6                              Children & Family Movies   
7      Dramas, Independent Movies, International Movies   
...                                                 ...   
8801            Dramas, International Movies, Thrillers   
8802                     Cult Movies, Dramas, Thrillers   
8804                            Comedies, Horror Movies   
8805                 Children & Family Movies, Comedies   
8806     Dramas, International Movies, Music & Musi

The first solution uses .drop with axis=0 to drop a row. 
The second identifies the empty values and takes the non-empty values by using the negation operator ~ while the third solution uses .dropna to drop empty rows within a column.

If you want to save the output after dropping, use inplace=True as a parameter. In this simple example, we’ll not drop any column or row.

#### Replace it by the mean,median or mode
Another common approach is to use the mean, median or mode to replace the empty values.the mean and median are used to replace numeric data,while the mode replaces categorical data.

As we’ve seen before, the rating column contains 0.16% of missing data. We could easily complete that tiny portion of data with the mode since the rating is a categorical value.


In [18]:
mode = ''.join(df_netflix_2019['rating'].mode())
df_netflix_2019['rating'].fillna(mode, inplace=True)

First ,we calculated the mode(TV-MA), and then we filled all the empty values with .fillna


#### Replace it by an arbitrary number with .fillna()

If the data is numeric, we can also set an arbitrary number to prevent removing any row without affecting our model's results.

If the duration column was a numeric value (currently, the format is string e.g. 90 minutes), we could replace the empty values by 0 with the following code.

In [17]:
df_netflix_2019['duration'].fillna(0, inplace=True)

Also, you can use the ffill , bfill to propagate the last valid observation forward and backward, respectively. This is extremely useful for some datasets but it’s not useful in the df_netflix_2019 dataset.

### 4.Identifying Outliers
An outlier is that data that differs significantly from other observations.A dataset might contain real outliers or outliers obtaines after poor data collection or caused by data entry errors

#### Using histograms to identify outliers within numeric data
We’re going to use the duration as a reference that will help us identify outliers in the Netflix catalog. The duration column is not considered a numerical value (e.g., 90) in our dataset because it’s mixed with strings (e.g., 90 min). Also, the duration of TV shows is in seasons (e.g., 2 seasons) so we need to filter it out.

With the following code, we’ll take only movies from the dataset and then extract the numeric values from the duration column.

In [22]:
#Creating column(extract)
df_movie = df_netflix_2019[df_netflix_2019['type']=='Movie']
df_movie = df_movie.assign(minute = df_movie['duration'].str.extract(r'(\d+)', expand=False).astype(int))

ValueError: cannot convert float NaN to integer

In [23]:
df1 = df_netflix_2019['duration']
df1

0          90 min
1       2 Seasons
2        1 Season
3        1 Season
4       2 Seasons
          ...    
8802      158 min
8803    2 Seasons
8804       88 min
8805       88 min
8806      111 min
Name: duration, Length: 8807, dtype: object

In [29]:
df_movie = df_netflix_2019[df_netflix_2019['type']=='Movie']
df_movie

show_id   type                             title  \
0         s1  Movie              Dick Johnson Is Dead   
6         s7  Movie  My Little Pony: A New Generation   
7         s8  Movie                           Sankofa   
9        s10  Movie                      The Starling   
12       s13  Movie                      Je Suis Karl   
...      ...    ...                               ...   
8801   s8802  Movie                           Zinzana   
8802   s8803  Movie                            Zodiac   
8804   s8805  Movie                        Zombieland   
8805   s8806  Movie                              Zoom   
8806   s8807  Movie                            Zubaan   

                           director  \
0                   Kirsten Johnson   
6     Robert Cullen, José Luis Ucha   
7                      Haile Gerima   
9                    Theodore Melfi   
12              Christian Schwochow   
...                             ...   
8801                Majid Al Ansari   
8802                  David Fincher   
8804                Ruben Fleischer   
8805                   Peter Hewitt   
8806                    Mozez Singh   

                                                   cast  \
0                                                   NaN   
6     Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
7     Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
9     Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
12    Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   
...                                                 ...   
8801  Ali Suliman, Saleh Bakri, Yasa, Ali Al-Jabri, ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...   

                                                country          date_added  \
0                                         United States  September 25, 2021   
6                                                   NaN  September 24, 2021   
7     United States, Ghana, Burkina Faso, United Kin...  September 24, 2021   
9                                         United States  September 24, 2021   
12                              Germany, Czech Republic  September 23, 2021   
...                                                 ...                 ...   
8801                       United Arab Emirates, Jordan       March 9, 2016   
8802                                      United States   November 20, 2019   
8804                                      United States    November 1, 2019   
8805                                      United States    January 11, 2020   
8806                                              India       March 2, 2019   

      release_year rating duration  \
0             2020  PG-13   90 min   
6             2021     PG   91 min   
7             1993  TV-MA  125 min   
9             2021  PG-13  104 min   
12            2021  TV-MA  127 min   
...            ...    ...      ...   
8801          2015  TV-MA   96 min   
8802          2007      R  158 min   
8804          2009      R   88 min   
8805          2006     PG   88 min   
8806          2015  TV-14  111 min   

                                             listed_in  \
0                                        Documentaries   
6                             Children & Family Movies   
7     Dramas, Independent Movies, International Movies   
9                                     Comedies, Dramas   
12                        Dramas, International Movies   
...                                                ...   
8801           Dramas, International Movies, Thrillers   
8802                    Cult Movies, Dramas, Thrillers   
8804                           Comedies, Horror Movies   
8805                Children & Family Movies, Comedies   
8806    Dramas, International Movies, Music & Musicals   

                                       

Now the data is ready to be displayed in a histogram.you can make plots with matplotlib, seaborn or pandas in python. In this case, I'll do it with matplotlib